src/uct/tcp/tcp_ep.c

```diff
@@ -658,7 +736,55 @@ uct_tcp_ep_comp_recv_am(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep,
     uct_iface_invoke_am(&iface->super, hdr->am_id, hdr + 1, hdr->length, 0);
 }
 
-unsigned uct_tcp_ep_progress_rx(uct_tcp_ep_t *ep)
+static inline ucs_status_t
+uct_tcp_ep_put_rx_advance(uct_tcp_ep_t *ep, uct_tcp_ep_put_hdr_t *put_req,
+                          size_t recv_length)
+{
+    put_req->addr   += recv_length;
+    put_req->length -= recv_length;
+
+    if (!put_req->length) {
+        uct_tcp_ep_post_put_ack(ep);
+        if (ep->ctx_caps & UCS_BIT(UCT_TCP_EP_CTX_TYPE_PUT_RX)) {
```

Comment:
maybe better set `UCT_TCP_EP_CTX_TYPE_PUT_RX` in the beginning of `uct_tcp_ep_handle_put_req`? So this func is always called with PUT_RX context. Then you would not need this branch (which is not performance critical here, but this would  look simpler overall)

Reply:
when handling PUT REQ in `uct_tcp_ep_handle_put_req` we don't want to release RX AM buffer as it may have other AMs
So, we can't set `UCT_TCP_EP_CTX_TYPE_PUT_RX` flag since it will always release RX AM buffer if PUT payload is fully received to RX AM buffer